### ISSUE TITLE
Add status information to the ServingRuntime CRD instances

### ISSUE THREAD

**Issue by bherta** [kserve/modelmesh-serving#213]
**Problem description:**
Today each instance of a `ServingRuntime` CRD contains a spec, but no status.  If I want to see the current status of an active `ServingRuntime`, such as to see if the number of pods matches the value I specified in the `replicas:` field, I need to look at the corresponding `Deployment`.  This isn't difficult to do manually, but in an environment monitored by [ArgoCD health checks](https://argo-cd.readthedocs.io/en/stable/operator-manual/health/), it prevents a problem that's visible at the lower Deployment level from getting surfaced at the top level, because even though the Deployment object may have a problem, it is associated with a ServingRuntime object that does not show any health issues, and so the overall application shows as healthy.

**Proposed Solution:**
We propose that status information is added to the ServingRuntime CRD, so that the health of a ServingRuntime instance in the cluster reflects the health of the corresponding Deployment (if one exists), with minimal status information such as the number of replicas and overall health.  This could be implemented by extending the ServingRuntime CRD, and adding code into the `modelmesh-controller` to update the status based on the corresponding Deployment.

**Alternatives:**
ArgoCD [provides a mechanism](https://argo-cd.readthedocs.io/en/stable/operator-manual/health/#way-1-define-a-custom-health-check-in-argocd-cm-configmap) to add custom health checks for any arbitrary CRD, but this mechanism only enables mapping values from fields available within the object instance.  If there is no status field in the object, then there is nothing we can use to reflect the health of the CRD instance.  We have not found a mechanism that enables ArgoCD to set the health status of an object based on the status of a different object, in this case the Deployment.

**Comment by njhill** [comment]
Hi @bherta thanks a lot for the issue writeup. I think we'd intentionally left out status from the `ServingRuntime` CRD, since each SR is intended to be more of a template than logically associated with specific instantiations of the runtime in question.

It happens that with ModelMesh Serving, each SR currently corresponds to a single deployment, but this isn't the case with KServe "classic" where the same SR CRD corresponds to a per-InferenceService deployment or Knative Service, i.e. one-to-many. It also won't be the case once `ClusterServingRuntime`s are supported on the modelmesh side (soon, @chinhuang007 is looking at this now).

That said, I understand the problem/desire ... maybe we can come up with something generic enough to cover all of these cases? This could be tricky for the `ClusterServingRuntime` case though, since I imagine all users may have get/list permission for those but shouldn't be able to see their status within namespaces that the user doesn't have access to.

**Comment by lizzzcai** [comment]
Hi @njhill , we also have the requirement for the status in the ServingRuntime. In our use case, we will have a lot of custom servingRuntime, and the user can get/create/patch/delete it, as our user mostly using custom frameworks with their own logic.

Currently, once a ServingRuntime is created, it creates a deployment but scales to zero. Users don't know whether their servingRuntime is working functionally until the deployment is scaled up. We will expect that once the servingRuntime is created, the deployment will be scaled at least 1 to ensure it is runable then scale back to zero to save cost and then update the status in the servingRuntime that this servingRuntime is running. Users should know whether the servingruntime is runable or not rather than facing the error when they try to deploy a model to the servingruntime.

**Comment by bherta** [comment]
@njhill Sorry for the delay in coming back to this.  The basic issue we have right now is that in the metadata section of the deployment object created by the controller for a serving runtime, the [ownerReferences field](https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/#owner-references-in-object-specifications) points to the ServingRuntime instance for that model type.  ArgoCD figures out the overall health of an application by looking at objects it directly controls (like the ServingRuntime and model-mesh controller), or by following the ownerReferences for dynamically created objects (like the deployment).  ArgoCD can't directly monitor the deployment that it doesn't create, so the only way to make this work today is either to change the ownerReference of the deployment to something that ArgoCD controls, or to add the status information of the associated deployment into its ServingRuntime instance.

I don't think it makes sense to change the ownerReference to something else.   The only candidate I can think of is the controller, but if one of many ServingRuntimes is unhealthy that doesn't mean that the controller is unhealthy.  It may technically solve this issue, but doesn't feel right, and it has the same problem as the ClusterServingRuntime, in that there may be multiple users with their own deployments that should not care about the status of deployments belonging to other users.  It might be possible to create a new CRD independent of the ServingRuntime that does nothing other than reflect the hearth status of the deployments, such as a "ServingRuntimeHealth" but splitting that role out from the ServingRuntime also feels unnecessary. So that leaves adding status information to the ServingRuntimes.

In the model-mesh implementation, the ServingRuntime status would be healthy if either there is no associated deployment, or the associated deployment is healthy.

Regarding the classic kserve use case, with multiple deployments for a single ServingRuntime, I think this is straightforward too.  The overall status of the ServingRuntime would be healthy, as long as there are no associated deployments, or all of the associated deployments are healthy.  This is the same pattern as the health of a deployment reflecting the health of all the replicas within it.

As for the ClusterServingRuntime, I don't think there is a good solution, but I don't think that we necessarily need one.  The ClusterServingRuntime doesn't have a status today, and it's a different CRD from the ServingRuntime.  Adding a status to the ServingRuntime won't change the existing behavior of the ClusterServingRuntime, which would continue as it is today.  If for whatever reason someone is in a similar situation with ArgoCD or some other utility that follows the ownerReferences metadata to collect the overall health of the application, and they are using the ClusterServingRuntime, they can always add the corresponding ServingRuntimes into their namespace, to provide that ownership reference and a way to bubble up the status.

We are also pursuing a solution on the ArgoCD side (see https://github.com/argoproj/argo-cd/issues/10590), but that may or may not go anywhere.

**Comment by njhill** [comment]
Thanks @bherta, I agree with all of the above. I think I'm good with adding status to ServingRuntimes, and you make a great point that ClusterServingRuntimes don't need to share the same status type.

We may want to think a bit about how to structure it / what fields to include. For example it would probably make sense to include running a replica count. We could consider including predictor counts too though this would be a bit more complicated because of cases where there are multiple runtimes on which a given predictor could run.

For unhealthy runtimes, it would be good to have states to differentiate between spec validation errors and runtime deployment errors, and having some kind of message field that could be populated with some detail about the error in question (transferred from deployment / replicaset / pod statuses in latter case).

---

@lizzzcai your comments about initial verification of new runtimes also sound interesting to me. Probably we should consider that as a separate / follow-on enhancement but should keep it in mind when thinking of the ServingRuntime state flows.

My thoughts implementation off this: The controller can maintain an annotation on the deployment associated with a SR whose value is a hash of the SR's spec and which is only set once a pod for the deployment has been verified to run successfully. If the hash is missing or doesn't match, and the deployment would otherwise be configured with 0 replicas, the controller would set it to 1 replica.

**Comment by bherta** [comment]
@njhill I'd like to propose the following status for ServingRuntimes:

```
status:
  replicas: 2
  availableReplicas: 2
  conditions:
    - type: Ready
      status: 'True'
      lastUpdateTime: '2022-10-21T14:08:43Z'
      lastTransitionTime: '2022-10-21T14:08:43Z'
      reason: MinimumReplicasAvailable
      message: Deployment has minimum availability.
```

This status is a blend of information related to the `ServingRuntime` spec, the associated deployment, and the status format of the `InferenceService`.

`replicas` & `availableReplicas`: These values are taken from the associated Deployment status field of the same names, if there are any `InferenceService` instances running on this `ServingRuntime`.  If the `Deployment` should not exist (e.g. because there are not `InferenceService` instances using it and the `ServingRuntime` is configured to ScaleToZero), `replicas` is set to 0.  Replicas represents the number of pods that should exist, and `availableReplicas` represents the actual number of pods that do exist.  Both values may be different from the replicas field in the spec, so I think they both belong in the spec as top-level fields.

'Ready' condition type: This contains the same information as the `Deployment`'s `Available` condition.  We use `Ready` rather than `Available` here because `Ready` is more standard, and is this is the name used in the existing `InferenceService` status.  If the `Deployment` should not exist, status is `True`, with a reason of `ScaledToZero` and message "No InferenceService instances for this ServingRuntime".  Otherwise the values are just copied from the associated `Deployment`.

There are a lot more conditions that we *could* add to this, such as more detail from the associated `Deployment` and `InferenceService` instances when they exist.  I'm avoiding these for now, mostly because they are not necessary for @lizzzcai and my original use cases to work.  More conditions can always be added in the future if needed, but also the `Deployment` and `InferenceServices` have their own status that people can look at directly if they need that information, so I'm not convinced it should be copied here.

I'm also taking the position that a `ServingRuntime` is healthy if the associated `Deployment` is healthy or should not exist.  In other words, the status of the `ServingRuntime` is good if it is in a state to handle InferenceService instances.  As for the case where an `InferenceService` is not working correctly (e.g. the puller can not find the model binary in storage), this doesn't necessarily mean there is any problem with the `ServingRuntime`, so the `Ready` condition can be `True` even when `InferenceServices` are not healthy.

**Comment by gaodan-fang** [comment]
Hi @njhill 
>I'm still wondering about this, because the separate deployments/knative services really correspond to the InferenceService in this case. The ServingRuntime is really more just like the template that they use and errors are just as (more?) likely to be related to problems with the specific predictor etc. So having the health of the SR be a function of the health of the IS deployments that use it would be kind of misleading I think.

IIRC, InferenceService failure will not lead to ServingRuntime's deployment failure (SR's deployment status won't reflect the status of IS). IS represents models need to be loaded into Serving Runtimes. One SR's deployment could have multiple InferenceServices, some of them could be loaded correctly, some of them may be not due to those and these errors, in such case, the SR's deployment's will still be healthy overall and the IS's status will reflect it's status correctly.

**Comment by bherta** [comment]
> I do think we would want to expose a bit more information though, in particular the name of the corresponding deployment if that's what they need to check in the unhealthy case, and some message indicating spec validation errors if the controller is blocked from processing the current spec for that reason.

Agreed on having the name of the deployment available in the status information.  This same situation occurs with Deployments containing the name of their associated Replicaset.  In that case, the name of the replicaset is identified in the message field of the deployment.  For example: `message: ReplicaSet "modelmesh-serving-kserve-mm-intent-5847c4c489" has successfully progressed.` (see a larger example [here](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#failed-deployment)).  This doesn't change my proposed status structure, it just improves the message field.

If there is spec validation error, this is something a user would see when trying to create the CRD instance, and the object won't actually be created in the cluster.  Is that correct?  If so, then there isn't any ability to put invalid spec errors in the status.  Unless you're referring to the spec of some other associated CRD?

> I'm still wondering about this, because the separate deployments/knative services really correspond to the InferenceService in this case. The ServingRuntime is really more just like the template that they use and errors are just as (more?) likely to be related to problems with the specific predictor etc. So having the health of the SR be a function of the health of the IS deployments that use it would be kind of misleading I think.

I don't think there is a need to treat the classic non-model-mesh use case (each InferenceService gets its own deployment) any differently from the model-mesh use case (multiple InferenceServices share a deployment).  In both of these, there is a bottom layer, which is the runtime pod(s), and then a layer on top of that for the InferenceService.  Getting the bottom layer up and running is a matter of allocating resources, pulling the docker image, etc, and its status is a reflection of if these things went ok.  Once that layer is in a useable state, then the InferenceService can be started, which involves pulling in model binaries, loading them into the runtime, etc, and these are the things reflected in the InferenceService status.  In both the model-mesh and classic scenarios, it's possible for the InferenceServices to have problems (e.g. permission issues pulling models from storage), but the Deployment(s) underneath them to be healthy (containers are running).  So from my perspective, the status of the ServingRuntime should represent the status of all the associated deployments in both the classic and model-mesh use cases, and the status of the InferenceServices doesn't need to be added today.

> It would be meaningful in both cases to include the count of InferenceServices using that SR (and maybe healthy/unhealthy counts of those), but this number is orthogonal/additional to the proposed replica counts.

This is also an example where the flexibility of Conditions can be useful.  If we ever decide we want to be able to see more status information about the InferenceServices in the ServingRuntime, and not just the deployments, it's easy to add more conditions without officially changing the spec.  I don't technically need these for my use case, so I would prefer to implement just the deployment information in the status today so that it can go in more quickly, and then we can always add more details about the InferenceServices in future updates.

**Comment by lizzzcai** [comment]
I agree with @gaodan-fang and @bherta on some of the points and the status format.

servingruntime can be running fine but the inference service (isvc) may fail to load due to different reasons. However, I think the servingruntime doesn't need to care about the status of the `isvc` running on top of it (maybe can add the count of healthy `isvc` etc). For isvc, maybe can show the status of the servingruntime in its status (if the overhead is less.). 

> The ServingRuntime is really more just like the template that they use and errors are just as (more?) likely to be related to problems with the specific predictor etc. So having the health of the SR be a function of the health of the IS deployments that use it would be kind of misleading I think.

> I don't think there is a need to treat the classic non-model-mesh use case (each InferenceService gets its own deployment) any differently from the model-mesh use case (multiple InferenceServices share a deployment).

I do think currently we can treat the classic `isvc` (inference service construct the main knative service + clusterServingRuntime provides the template, most of the model format in kserve using clusterServingRuntime) and the modelmesh `isvc` (inference service provides the model artifacts + [cluster]ServingRuntime provide the running deploymentHost, most of the runtime in modelmesh using namespace-scoped servingruntime) `differently`. 
As there are two different controllers for these two use cases and one mainly uses clusterServingRuntime and another one uses ServingRuntime. In both cases, ClusterServingRuntime probably can only be treated as a template (having status in a cluster-scope resource maybe not a good idea here).

For me I will prefer to have minimal status information like the example @bherta shown for  namespace-scoped ServingRuntime, and have an `initial-scale: 1` for servingRuntime deployment at the beginning to initialize the status

**Comment by njhill** [comment]
Thanks @bherta @gaodan-fang @lizzzcai 

> If there is spec validation error, this is something a user would see when trying to create the CRD instance, and the object won't actually be created in the cluster. Is that correct?

We don't actually use validation hooks currently since there is extra complication in configuring/securing the webhook endpoint, so nothing would prevent the CR from being created. But maybe that's something to look at since I think they are used in the main KServe controller. But yes probably reasonable to not worry about them in the status.

> In both the model-mesh and classic scenarios, it's possible for the InferenceServices to have problems (e.g. permission issues pulling models from storage), but the Deployment(s) underneath them to be healthy (containers are running). So from my perspective, the status of the ServingRuntime should represent the status of all the associated deployments in both the classic and model-mesh use cases, and the status of the InferenceServices doesn't need to be there for it to be useful.

I'm not so sure about that though. How do you disentangle whether a particular single-model InferenceService deployment (or KNative Service) is failing due to a underlying "runtime"  issue or an issue with the specific model being that it's loading (or config of that specific IS)? From Kubernetes' point of view it just dies or does not become ready.

> one mainly uses clusterServingRuntime and another one uses ServingRuntime.

I think this is more of a point-in-time statement since ClusterServingRuntime support was added later on the ModelMesh-Serving side, but it's there now and the built-in multi-model runtimes in the next release will be ClusterServingRuntimes alongside the other KServe ones. The non-Cluster ServingRuntimes I think also apply equally to both and are really for those that you want to create/manage at a namespace level, or for use with namespace-scope controller mode installs (where you could have independent ModelMesh-Servings instances in different namespaces).

> In both cases, ClusterServingRuntime probably can only be treated as a template (having status in a cluster-scope resource maybe not a good idea here).

Yes I think we're all in agreement on that!

> This is also an example where the flexibility of Conditions can be useful. If we ever decide we want to be able to see more status information about the InferenceServices in the ServingRuntime, and not just the deployments, it's easy to add more conditions without officially changing the spec.

But Conditions reflect binary (or ternary) states, I'm not sure how you would use that to reflect a count of active InferenceServices using a particular ServingRuntime for example. I know there have been long debates over the merits of Conditions, I thought the most recent conclusion was that they weren't a good idea (without digging through issue history in the Kube repo, there is for example [this](https://github.com/kubernetes-sigs/cluster-api/blob/112951ee/docs/proposals/20181121-machine-api.md#conditions) reference). And new fields can be added to the Status object in a backwards-compatible way.

But again if everyone wants to use Conditions I'm ok with it. 

> I don't technically need these for my use case, so I would prefer to implement just the deployment information in the status today so that it can go in more quickly, and then we can always add more details about the InferenceServices in future updates.

Fair enough, this is fine with me!

> For me I will prefer to have minimal status information like the example @bherta shown for namespace-scoped ServingRuntime, and have an `initial-scale: 1` for servingRuntime deployment at the beginning to initialize the status (fail earlier).

As mentioned in my [prior comment](https://github.com/kserve/modelmesh-serving/issues/213#issuecomment-1258389187) I like the idea of this but it's probably best to do as a follow-on enhancement.

**Comment by lizzzcai** [comment]
> I'm not so sure about that though. How do you disentangle whether a particular single-model InferenceService deployment (or KNative Service) is failing due to a underlying "runtime" issue or an issue with the specific model being that it's loading (or config of that specific IS)? From Kubernetes' point of view it just dies or does not become ready.

In the single-model use case, user mainly uses storage-initializer (init-container) to download the model. If the error is in the model pulling stage, the pod will fail in the init container stage. user is able to check the error in the inferenceService status, logs from init container or k8s events. After that, the main container will start (check pod status if there are image pulling errors) and run (Check the logs of the pod. It is possible to load the model only when the first request comes for custom runtime). 

> I think this is more of a point-in-time statement since ClusterServingRuntime support was added later on the ModelMesh-Serving side, but it's there now and the built-in multi-model runtimes in the next release will be ClusterServingRuntimes alongside the other KServe ones. The non-Cluster ServingRuntimes I think also apply equally to both and are really for those that you want to create/manage at a namespace level, or for use with namespace-scope controller mode installs (where you could have independent ModelMesh-Servings instances in different namespaces).

If cluster/ServingRuntime is a pure template concept, I feel something is missing to make it an instance. like:
```
ClusterServingRuntime[Template] -> ClusterServingRuntimeInstance (show status) -> deployment.
```
> As mentioned in my https://github.com/kserve/modelmesh-serving/issues/213#issuecomment-1258389187 I like the idea of this but it's probably best to do as a follow-on enhancement.

Sure, we can take it as a follow-on enhancement.
